samples/web/content/apprtc/room.py

```diff
@@ -65,8 +81,70 @@ def has_room(host, room_id):
   room = memcache_client.get(key)
   return room is not None
 
+def get_room(host, room_id):
+  key = get_memcache_key_for_room(host, room_id)
+  memcache_client = memcache.Client()
+  return memcache_client.get(key)
+
+def get_room_state(host, room_id):
+  room = get_room(host, room_id)
+  if room is None:
+    return None
+  return room.get_room_state()
+
+def remove_room_for_declined_call(host, room_id, callee_gcm_id):
+  memcache_client = memcache.Client()
+  key = get_memcache_key_for_room(host, room_id)
+  
+  retries = 0
+  while retries < constants.ROOM_MEMCACHE_RETRY_LIMIT:
+    room = memcache_client.gets(key)
+    
+    if room is None:
+      logging.warning('Can\'t remove room ' + room_id +
+          ' because it doesn\'t exist, client: ' + callee_gcm_id)
+      return constants.RESPONSE_INVALID_ROOM
+  
+    if not room.is_client_allowed(callee_gcm_id):
+      logging.warning('Can\'t remove room ' + room_id +
+          ' because room does not allow client ' + callee_gcm_id)
+      return constants.RESPONSE_INVALID_CALLEE
+
+    # The client already in the room is the caller.
+    # The caller should not be removing the room via decline.
+    if room.has_client(callee_gcm_id):
+      logging.warning('Can\'t remove room ' + room_id +
+          ' because client is caller: ' + callee_gcm_id)
+      return constants.RESPONSE_INVALID_CALLEE
+  
+    if room.room_type != Room.TYPE_DIRECT:
+      logging.warning('Can\'t remove room ' + room_id + 
+          ' because it has type: ' + str(room.room_type) + ' client: ' +
+          callee_gcm_id)
+      return constants.RESPONSE_INVALID_ROOM
+  
+    if room.get_room_state() == Room.STATE_FULL:
+      logging.warning('Can\'t remove room ' + room_id + 
+          'because it is full (' + str(room.get_occupancy()) +') client: ' +
+          callee_gcm_id)
+      return constants.RESPONSE_INVALID_ROOM
+  
+    # Reset the room to the initial state so it may be reused.
+    room.reset()
+    if memcache_client.cas(key, room, constants.ROOM_MEMCACHE_EXPIRATION_SEC):
+      logging.info('Reset room %s to base state in remove_room by client %s, retries = %d' \
+          %(room_id, callee_gcm_id, retries))
+      return constants.RESPONSE_SUCCESS
+    else:
+      retries = retries + 1
+    
+    logging.warning('Failed to remove room ' + room_id + ' after retry limit ' +
+        ' client: ' + callee_gcm_id)
+    return constants.RESPONSE_INTERNAL_ERROR
```

Comment:
Good eye. I also realized I should add return statements to a couple of the loops that were previously infinite loops.